src/form/Input.js

```diff
@@ -0,0 +1,159 @@
+    function Input() {
+        HTMLWidget.call(this);
+        this._class = "form_Input";
+
+        this._tag = "div";
+    }
+    Input.prototype = Object.create(HTMLWidget.prototype);
+    Input.prototype.publish("id", "", "string", "HTML id for the input");
+    Input.prototype.publish("label", "", "string", "Descriptive label");
+    Input.prototype.publish("type", "text", "set", "Input type", ["textbox","number","checkbox","button","select","textarea"]);
+    Input.prototype.publish("validate", null, "string", "Input type");
+    Input.prototype.publish("selectOptions", [], "array", "Array of options used to fill a dropdown list");
+    Input.prototype.publish("value", "", "string", "Input type");
+
+    Input.prototype.testData = function (widget) {
+        return this;
+    };
+    
+    Input.prototype.enter = function (domNode, element) {
+        HTMLWidget.prototype.enter.apply(this, arguments);
+        switch(this.type()){
+            case 'select':
+                this._inputElement = element.append("select");
+                break;
+            case 'textarea':
+                this._inputElement = element.append("textarea");
+                break;
+            case 'button':
+                this._inputElement = element.append("button");
+                break;
+            default:
+                this._inputElement = element.append("input").attr('type',this.type());
+                break;
+        }
+    };
+    Input.prototype.update = function (domNode, element) {
+        HTMLWidget.prototype.update.apply(this, arguments);
+        this._inputElement.attr('id', this.id());
+        
+        switch(this.type()){
+            case 'select':
+                this.checkNodeName('SELECT',element);
+                this._inputElement.attr('value', this.value());
+                this.insertSelectOptions(this.selectOptions());
+                break;
+            case 'textarea':
+                this.checkNodeName('TEXTAREA',element);
+                this._inputElement.node().innerHTML = this.value();
+                break;
+            case 'button':
+                this.checkNodeName('BUTTON',element);
+                this._inputElement.node().innerHTML = this.value();
+                break;
+            case 'checkbox':
+                this.checkNodeName('INPUT',element);
+                this._inputElement.property('checked', this.value());
+                break;
+            default:
+                this.checkNodeName('INPUT',element);
+                this._inputElement.attr('value', this.value());
+                this._inputElement.attr('type', this.type());
+                break;
+        }
+        if(typeof(this._blur) !== 'function'){
+            this._blur = function(){};
+        }
+        if(typeof(this._click) !== 'function'){
+            this._click = function(){};
+        }
+        if(typeof(this._change) !== 'function'){
+            this._change = function(){};
+        }
```

Comment:
why is the typeof and adding of default necessary here can this be moved below into the .blue and .click and .change functions?